Mathlib/Data/List/TakeDrop.lean

```diff
@@ -150,6 +150,69 @@ theorem span_eq_takeWhile_dropWhile (l : List α) : span p l = (takeWhile p l, d
 
 end Filter
 
+section Splits
+
+/-! ### Generate splits of a list. -/
+
+/-- Pairs of prefixes and suffixes of a list. -/
+def splits (xs : List α) : List (List α × List α) :=
+  List.map xs.splitAt (range (xs.length + 1))
+
+@[simp]
+lemma splits_nil : (List.nil (α:=α)).splits= [([], [])] := by rfl
+
+@[simp]
+lemma splits_cons (x : α) (xs : List α) :
+    List.splits (x :: xs) =
+    ([], x :: xs) :: List.map (fun td ↦ (x :: td.1, td.2)) (List.splits xs) := by
+  simp [List.splits, List.range_succ_eq_map]
+
+lemma splits_spec (x y l : List α) :
+    (x, y) ∈ l.splits ↔ l = x ++ y := by
+  simp [List.splits]
+  constructor
+  · rintro ⟨n, hn, htake, hdrop⟩
+    rw [←htake, ←hdrop, List.take_append_drop]
+  · rintro rfl
+    exists x.length
+    constructor
+    · simp; omega
+    · constructor
+      · rw [List.take_left]
+      · rw [List.drop_left]
+
+lemma splits_non_empty (l : List α) : ∃ td, td ∈ l.splits := by
+  exists ([], l)
+  simp [List.splits_spec]
+
+/-- Left-associative triple splits of a list. -/
+def splits3_l : List α → List (List α × List α × List α) :=
```

Comment:
Do we need the triple splits? It might be easier, in cases where something like this is needed, to just deal with two applications of the basic splits function and use the API there, this way we don't have to think about quadruple splits and API for triple splits and so on.